File: wiki-java/src/main/java/dev/freya02/wiki/SpringBot.java
package dev.freya02.wiki;

import io.github.freya022.botcommands.api.core.JDAService;
import io.github.freya022.botcommands.api.core.config.JDAConfiguration;
import io.github.freya022.botcommands.api.core.events.BReadyEvent;
import net.dv8tion.jda.api.entities.Activity;
import net.dv8tion.jda.api.hooks.IEventManager;
import net.dv8tion.jda.api.requests.GatewayIntent;
import net.dv8tion.jda.api.utils.cache.CacheFlag;
import org.jspecify.annotations.NullMarked;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.util.Set;

// --8<-- [start:jdaservice-java]
@Service
@NullMarked // Everything is non-null unless @Nullable
public class SpringBot extends JDAService {
    private final JDAConfiguration jdaConfiguration;
    private final String token;

    public SpringBot(JDAConfiguration jdaConfiguration, @Value("${bot.token}") String token) {
        this.jdaConfiguration = jdaConfiguration;
        this.token = token;
    }

    @Override
    public Set<CacheFlag> getCacheFlags() {
        return jdaConfiguration.getCacheFlags();
    }

    @Override
    public Set<GatewayIntent> getIntents() {
        return jdaConfiguration.getIntents();
    }

    @Override
    public void createJDA(BReadyEvent event, IEventManager eventManager) {
        // This uses JDABuilder#createLight, with the intents and the additional cache flags set above
        // It also sets the EventManager and a special rate limiter
        createLight(token)
                .setActivity(Activity.customStatus("In Java with ❤️"))
                .build();
    }
}
// --8<-- [end:jdaservice-java]

File: wiki-kotlin-commons/src/main/kotlin/dev/freya02/wiki/SpringBot.kt
package dev.freya02.wiki

import io.github.freya022.botcommands.api.core.JDAService
import io.github.freya022.botcommands.api.core.config.JDAConfiguration
import io.github.freya022.botcommands.api.core.events.BReadyEvent
import io.github.freya022.botcommands.api.core.light
import net.dv8tion.jda.api.entities.Activity
import net.dv8tion.jda.api.hooks.IEventManager
import net.dv8tion.jda.api.requests.GatewayIntent
import net.dv8tion.jda.api.utils.cache.CacheFlag
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service

// --8<-- [start:jdaservice-kotlin]
@Service
class SpringBot(
    private val jdaConfiguration: JDAConfiguration,
    @param:Value($$"${bot.token}")
    private val token: String,
) : JDAService() {
    override val intents: Set<GatewayIntent> get() = jdaConfiguration.intents

    override val cacheFlags: Set<CacheFlag> get() = jdaConfiguration.cacheFlags

    override fun createJDA(event: BReadyEvent, eventManager: IEventManager) {
        // This uses JDABuilder#createLight, with the intents and the additional cache flags set above
        // It also sets the EventManager and a special rate limiter
        light(
            token,
            activity = Activity.customStatus("In Kotlin with ❤️")
        ) {
            // Other configs
        }
    }
}
// --8<-- [end:jdaservice-kotlin]
